File: backend/src/schemas/scenarios.py
"""Pydantic schemas for scenarios endpoints.

Covers: ScenarioCreate, ScenarioUpdate, ScenarioResponse,
ScenarioDetailResponse, ScenarioMACCDataResponse, CompareResponse,
AddInitiativesRequest, ReorderRequest
— per scenarios-api.md contract.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Initiative item inside a scenario detail
# ---------------------------------------------------------------------------


class ScenarioInitiativeItem(BaseModel):
    """A single initiative as it appears inside a scenario detail response."""

    id: str
    name: str
    capex_gbp: float
    opex_annual_gbp: float | None = None
    co2e_reduction_annual_tonnes: float
    cost_per_tonne: float
    lifespan_years: int
    status: str
    confidence: str | None = None
    display_order: int
    is_included: bool

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Target alignment info
# ---------------------------------------------------------------------------


class TargetAlignmentEntry(BaseModel):
    id: str
    target_year: int
    target_type: str
    target_value_pct: float
    baseline_co2e_tonnes: float
    target_co2e_tonnes: float | None = None
    projected_co2e_tonnes: float
    gap_co2e_tonnes: float
    on_track: bool


class TargetAlignmentInfo(BaseModel):
    has_targets: bool
    targets: list[TargetAlignmentEntry] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Scenario Create / Update
# ---------------------------------------------------------------------------


class ScenarioCreate(BaseModel):
    """Body for POST /scenarios."""

    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    is_baseline: bool = False
    initiative_ids: list[str] = Field(default_factory=list)


class ScenarioUpdate(BaseModel):
    """Body for PUT /scenarios/{id} — all fields optional."""

    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    is_baseline: bool | None = None


# ---------------------------------------------------------------------------
# Scenario responses
# ---------------------------------------------------------------------------


class ScenarioResponse(BaseModel):
    """Summary scenario response (list items and create/update responses)."""

    id: str
    name: str
    description: str | None = None
    is_baseline: bool
    total_capex_gbp: float
    total_opex_annual_gbp: float
    total_co2e_reduction_annual_tonnes: float
    residual_co2e_tonnes: float
    weighted_avg_cost_per_tonne: float
    initiative_count: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ScenarioDetailResponse(ScenarioResponse):
    """Full scenario with initiatives and target alignment."""

    initiatives: list[ScenarioInitiativeItem] = Field(default_factory=list)
    target_alignment: TargetAlignmentInfo


class ScenarioListResponse(BaseModel):
    items: list[ScenarioResponse]
    total: int


# ---------------------------------------------------------------------------
# Add / Reorder initiatives
# ---------------------------------------------------------------------------


class AddInitiativesRequest(BaseModel):
    initiative_ids: list[str] = Field(..., min_length=1)
    display_order_start: int = Field(1, ge=1)


class InitiativeOrderItem(BaseModel):
    initiative_id: str
    display_order: int


class ReorderRequest(BaseModel):
    initiative_order: list[InitiativeOrderItem] = Field(..., min_length=1)


# ---------------------------------------------------------------------------
# MACC data
# ---------------------------------------------------------------------------


class ScenarioMACCBar(BaseModel):
    initiative_id: str
    initiative_name: str
    width: float
    height: float
    x_start: float
    cost_per_tonne: float
    co2e_reduction_annual_tonnes: float
    capex_gbp: float
    opex_annual_gbp: float | None = None
    status: str
    confidence: str | None = None
    is_negative_cost: bool


class ScenarioMACCSummary(BaseModel):
    total_capex_gbp: float
    total_opex_annual_gbp: float
    net_negative_cost_initiatives: int
    net_positive_cost_initiatives: int
    weighted_avg_cost_per_tonne: float


class ScenarioMACCDataResponse(BaseModel):
    scenario_id: str
    scenario_name: str
    total_emissions_co2e_tonnes: float
    total_reduction_co2e_tonnes: float
    bars: list[ScenarioMACCBar]
    summary: ScenarioMACCSummary


# ---------------------------------------------------------------------------
# Compare
# ---------------------------------------------------------------------------


class ScenarioCompareItem(BaseModel):
    id: str
    name: str
    is_baseline: bool
    total_capex_gbp: float
    total_opex_annual_gbp: float
    total_co2e_reduction_annual_tonnes: float
    residual_co2e_tonnes: float
    initiative_count: int
    avg_cost_per_tonne: float
    meets_targets: bool
    gap_to_target_pct: float


class SharedInitiativeItem(BaseModel):
    id: str
    name: str
    in_scenarios: list[str]


class UniqueInitiativeItem(BaseModel):
    id: str
    name: str


class CompareResponse(BaseModel):
    scenarios: list[ScenarioCompareItem]
    shared_initiatives: list[SharedInitiativeItem]
    unique_initiatives: dict[str, list[UniqueInitiativeItem]]

File: frontend/src/types/scenarios.ts
/**
 * TypeScript types for context, targets, and scenarios.
 * Mirrors backend schemas: context.py, scenarios.py
 * and contracts: context-api.md, scenarios-api.md.
 */

// ---------------------------------------------------------------------------
// Organisational Context
// ---------------------------------------------------------------------------

export type SustainabilityMaturity = "beginner" | "intermediate" | "advanced";
export type TargetType = "absolute" | "intensity";

export interface OrgContext {
  id: string;
  organisation_id: string;
  industry_sector: string | null;
  employee_count: number | null;
  annual_revenue_gbp: number | null;
  operating_geographies: string[] | null;
  sustainability_maturity: SustainabilityMaturity | null;
  budget_constraint_gbp: number | null;
  target_year: number | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface OrgContextUpsert {
  industry_sector?: string | null;
  employee_count?: number | null;
  annual_revenue_gbp?: number | null;
  operating_geographies?: string[] | null;
  sustainability_maturity?: SustainabilityMaturity | null;
  budget_constraint_gbp?: number | null;
  target_year?: number | null;
  notes?: string | null;
}

// ---------------------------------------------------------------------------
// Emission Targets
// ---------------------------------------------------------------------------

export interface Target {
  id: string;
  organisation_id: string;
  target_year: number;
  target_type: TargetType;
  target_value_pct: number;
  baseline_year: number;
  baseline_co2e_tonnes: number;
  target_co2e_tonnes: number | null;
  scope_coverage: number[] | null;
  source: string | null;
  notes: string | null;
  created_at: string;
}

export interface TargetCreate {
  target_year: number;
  target_type: TargetType;
  target_value_pct: number;
  baseline_year: number;
  baseline_co2e_tonnes: number;
  scope_coverage?: number[] | null;
  source?: string | null;
  notes?: string | null;
}

export interface TargetUpdate {
  target_year?: number;
  target_type?: TargetType;
  target_value_pct?: number;
  baseline_year?: number;
  baseline_co2e_tonnes?: number;
  scope_coverage?: number[] | null;
  source?: string | null;
  notes?: string | null;
}

export interface TargetListResponse {
  items: Target[];
  total: number;
}

export interface TargetProgress {
  target: Target;
  current_co2e_tonnes: number;
  scenario_reduction_co2e_tonnes: number;
  projected_co2e_tonnes: number;
  gap_co2e_tonnes: number;
  on_track: boolean;
  coverage_pct: number;
  years_remaining: number;
}

// ---------------------------------------------------------------------------
// Scenarios
// ---------------------------------------------------------------------------

export interface ScenarioInitiativeItem {
  id: string;
  name: string;
  capex_gbp: number;
  opex_annual_gbp: number | null;
  co2e_reduction_annual_tonnes: number;
  cost_per_tonne: number;
  lifespan_years: number;
  status: string;
  confidence: string | null;
  display_order: number;
  is_included: boolean;
}

export interface TargetAlignmentEntry {
  id: string;
  target_year: number;
  target_type: TargetType;
  target_value_pct: number;
  baseline_co2e_tonnes: number;
  target_co2e_tonnes: number | null;
  projected_co2e_tonnes: number;
  gap_co2e_tonnes: number;
  on_track: boolean;
}

export interface TargetAlignmentInfo {
  has_targets: boolean;
  targets: TargetAlignmentEntry[];
}

export interface Scenario {
  id: string;
  name: string;
  description: string | null;
  is_baseline: boolean;
  total_capex_gbp: number;
  total_opex_annual_gbp: number;
  total_co2e_reduction_annual_tonnes: number;
  residual_co2e_tonnes: number;
  weighted_avg_cost_per_tonne: number;
  initiative_count: number;
  created_at: string;
  updated_at: string;
}

export interface ScenarioDetail extends Scenario {
  initiatives: ScenarioInitiativeItem[];
  target_alignment: TargetAlignmentInfo;
}

export interface ScenarioListResponse {
  items: Scenario[];
  total: number;
}

export interface ScenarioCreate {
  name: string;
  description?: string | null;
  is_baseline?: boolean;
  initiative_ids?: string[];
}

export interface ScenarioUpdate {
  name?: string;
  description?: string | null;
  is_baseline?: boolean;
}

export interface AddInitiativesRequest {
  initiative_ids: string[];
  display_order_start?: number;
}

export interface InitiativeOrderItem {
  initiative_id: string;
  display_order: number;
}

export interface ReorderRequest {
  initiative_order: InitiativeOrderItem[];
}

// ---------------------------------------------------------------------------
// MACC data
// ---------------------------------------------------------------------------

export interface ScenarioMACCBar {
  initiative_id: string;
  initiative_name: string;
  width: number;
  height: number;
  x_start: number;
  cost_per_tonne: number;
  co2e_reduction_annual_tonnes: number;
  capex_gbp: number;
  opex_annual_gbp: number | null;
  status: string;
  confidence: string | null;
  is_negative_cost: boolean;
}

export interface ScenarioMACCSummary {
  total_capex_gbp: number;
  total_opex_annual_gbp: number;
  net_negative_cost_initiatives: number;
  net_positive_cost_initiatives: number;
  weighted_avg_cost_per_tonne: number;
}

export interface ScenarioMACCData {
  scenario_id: string;
  scenario_name: string;
  total_emissions_co2e_tonnes: number;
  total_reduction_co2e_tonnes: number;
  bars: ScenarioMACCBar[];
  summary: ScenarioMACCSummary;
}

// ---------------------------------------------------------------------------
// Compare
// ---------------------------------------------------------------------------

export interface ScenarioCompareItem {
  id: string;
  name: string;
  is_baseline: boolean;
  total_capex_gbp: number;
  total_opex_annual_gbp: number;
  total_co2e_reduction_annual_tonnes: number;
  residual_co2e_tonnes: number;
  initiative_count: number;
  avg_cost_per_tonne: number;
  meets_targets: boolean;
  gap_to_target_pct: number;
}

export interface SharedInitiativeItem {
  id: string;
  name: string;
  in_scenarios: string[];
}

export interface UniqueInitiativeItem {
  id: string;
  name: string;
}

export interface CompareResponse {
  scenarios: ScenarioCompareItem[];
  shared_initiatives: SharedInitiativeItem[];
  unique_initiatives: Record<string, UniqueInitiativeItem[]>;
}
